aiohttp/helpers.py

```diff
@@ -279,6 +329,246 @@ def basicauth_from_netrc(netrc_obj: Optional[netrc.netrc], host: str) -> BasicAu
+        #             +--|----|------|----|--||-----|--> group quoted or unquoted
+        #                |    |      |    |  ||     |
+        #                +----|------|----|--||-----|--> if quoted...
+        #                     +------|----|--||-----|--> anything but " or \
+        #                            +----|--||-----|--> escaped characters allowed
+        #                                 +--||-----|--> or can be empty string
+        #                                    ||     |
+        #                                    +|-----|--> if unquoted...
+        #                                     +-----|--> anything but , or <space>
+        #                                           +--> at least one char req'd
+    )
+
+    header_pairs = {}
+    for key, quoted_val, unquoted_val in pattern.findall(header):
+        val = quoted_val if quoted_val else unquoted_val
+        if val:
+            val = val.replace('\\"', '"')  # unescape any escaped quotes
+        header_pairs[key] = val
+
+    return header_pairs
+
+
+class DigestAuth(AuthBase):
+    """
+    HTTP digest authentication helper.
+
+    The work here is based off of
+    https://github.com/requests/requests/blob/v2.18.4/requests/auth.py.
+
+    Please also refer to RFC7616.
+    """
+
+    def __init__(
+        self,
+        login: str,
+        password: str,
+    ) -> None:
+        super().__init__(login, password)
+        self.ctx = DigestAuthContext()
+
+    def encode(self, method: str, url: URL, body: Any) -> str:
+        """Build digest header"""
+        self.ctx.init_thread()
+
+        if not self.ctx.handled_401:
+            return ""
+
+        if "realm" not in self.ctx.challenge:
+            raise client_exceptions.ClientError("Challenge is missing realm")
+
+        if "nonce" not in self.ctx.challenge:
+            raise client_exceptions.ClientError("Challenge is missing nonce")
+
+        realm: str = self.ctx.challenge.get("realm", "")
+        nonce: str = self.ctx.challenge.get("nonce", "")
+        qop_raw: str = self.ctx.challenge.get("qop", "")
+        algorithm: str = self.ctx.challenge.get("algorithm", "MD5").upper()
+        opaque: str = self.ctx.challenge.get("opaque", "")
+
+        qop: str = ""
+        if qop_raw:
+            qop_list = [q.strip() for q in qop_raw.split(",") if q.strip()]
+            valid_qops = {"auth", "auth-int"}.intersection(qop_list)
+            if not valid_qops:
+                raise client_exceptions.ClientError(
+                    f"Unsupported qop value(s): {qop_raw}"
+                )
+
+            qop = "auth-int" if "auth-int" in valid_qops else "auth"
+
+        if algorithm not in DigestFunctions:
+            return ""
+        hash_fn: Final = DigestFunctions[algorithm]
+
+        def H(x: str) -> str:
+            return hash_fn(x.encode()).hexdigest()
+
+        def KD(s: str, d: str) -> str:
+            return H(f"{s}:{d}")
```

Comment:
Please give these docstrings. and link to sections in the RFC that define that these are so anyone who comes across this code later will know what this is doing

https://datatracker.ietf.org/doc/html/rfc2617#section-3.2.2.2